### PR TITLE
G-API: ONNX. Support tensor input for CNN with dynamic input 

### PR DESCRIPTION
**Pull request by mpashchenkov**
### ONNX. Support tensor input for CNN with dynamic input 

- Changed assert for CV_32F case. Added test.

<cut/>

### Added new tests and refactored olds. 

New tests:
- InferTensor;
- InferDynamicInputTensor;
- Infer(Gray)Image;
- InferMultiNodes.

Refactored tests:
- Infer;
- InferROI;
- InferROIList;
- Infer2ROIList.

Tests contain unsupported(OCV bb) models:
- tinyyolov2-8.onnx from https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/tiny-yolov2

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom
buildworker:Custom=linux-1
build_image:Custom=ubuntu-onnx:20.04
test_modules:Custom=gapi
test_filter:Custom=*ONNX*
```

### CODE REVIEW

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 1]
```diff
@@ -12,172 +12,309 @@
...
+    cv::Mat processed_mat;
+    preprocess(in_mat1, processed_mat);
+    infer<float>(processed_mat, out_onnx.front());
     // G_API code
     G_API_NET(SqueezNet, <cv::GMat(cv::GMat)>, "squeeznet");
     cv::GMat in;
     cv::GMat out = cv::gapi::infer<SqueezNet>(in);
     cv::GComputation comp(cv::GIn(in), cv::GOut(out));
-    // NOTE: We have to normalize U8 tensor
-    // so cfgMeanStd() is here
```
Comment (OrestChura) [inline modules/gapi/test/infer/gapi_infer_onnx_test.cpp:156]: Isn't this comment still actual? Preserve, m b ?

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 2]
```diff
@@ -264,15 +381,82 @@ TEST(ONNX, Infer2ROIList)
...
+    const cv::Mat in_tensor = tensor / 255.f;
+    // ONNX_API code
+    infer<float>(in_tensor, out_onnx.front());
+    // G_API code
+    G_API_NET(YoloNet, <cv::GMat(cv::GMat)>, "YoloNet");
+    cv::GMat in;
+    cv::GMat out = cv::gapi::infer<YoloNet>(in);
+    cv::GComputation comp(cv::GIn(in), cv::GOut(out));
+    auto net = cv::gapi::onnx::Params<YoloNet>{model_path}
+        .cfgPostProc({cv::GMatDesc{CV_32F, {1,125,13,13}}}, remap_yolo)
```
Comment (OrestChura) [inline modules/gapi/test/infer/gapi_infer_onnx_test.cpp:404]: Spacies

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 3]
```diff
@@ -264,15 +381,82 @@ TEST(ONNX, Infer2ROIList)
...
 }
 
+TEST_F(ONNXtest, InferMultOutput)
+{
+    useModel("object_detection_segmentation/ssd-mobilenetv1/model/ssd_mobilenet_v1_10");
+    // ONNX_API code
+    const auto prep_mat = in_mat1.reshape(1, {1, in_mat1.rows, in_mat1.cols, in_mat1.channels()});
+    infer<uint8_t>({prep_mat}, out_onnx);
+    // G_API code
+    using OUT = std::tuple<cv::GMat, cv::GMat, cv::GMat, cv::GMat>;
```
Comment (OrestChura) [inline modules/gapi/test/infer/gapi_infer_onnx_test.cpp:441]: Let it be `GMobileNetOut`, for example. Or `GSSDOut`

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 4]
```diff
@@ -264,15 +381,82 @@ TEST(ONNX, Infer2ROIList)
...
+    cv::GMat in;
+    cv::GMat out1, out2, out3, out4;
+    std::tie(out1, out2, out3, out4) = cv::gapi::infer<MobileNet>(in);
+    cv::GComputation comp(cv::GIn(in), cv::GOut(out1, out2, out3, out4));
+    auto net = cv::gapi::onnx::Params<MobileNet>{model_path}
+        .cfgOutputLayers({"out1", "out2", "out3", "out4"})
+        .cfgPostProc({cv::GMatDesc{CV_32F, {1,100,4}},
+                      cv::GMatDesc{CV_32F, {1,100}},
+                      cv::GMatDesc{CV_32F, {1,100}},
+                      cv::GMatDesc{CV_32F, {1,1}}}, remap_ssd_ports);
```
Comment (OrestChura) [inline modules/gapi/test/infer/gapi_infer_onnx_test.cpp:452]: Spaces, please

**Review by OrestChura — COMMENTED** [review]
Great tests!

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp [hunk 5]
```diff
@@ -167,8 +167,16 @@ inline void preprocess(const cv::Mat& src,
...
-        GAPI_Assert(toORT(src.size) == ti.dims && "32F tensor dimensions should match with NN input");
-        GAPI_Assert(!ti.is_dynamic && "Dynamic inputs are not supported for this case");
+        const auto tensor_dims = toORT(src.size);
+        if (tensor_dims.size() == ti.dims.size()) {
+            for (size_t i = 0; i < ti.dims.size(); ++i) {
+                GAPI_Assert((ti.dims[i] == -1 || ti.dims[i] == tensor_dims[i]) &&
+                            "32F tensor dimensions should match with all non-dynamic NN input dimensions");
+            }
+        } else {
+            GAPI_Assert(false && "32F tensor size should match with NN input");
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:177]: We check size.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 6]
```diff
@@ -12,172 +12,309 @@
...
 
-    double normInf = cvtest::norm(ref, test, cv::NORM_INF);
+    const double normInf = cvtest::norm(ref, test, cv::NORM_INF);
     EXPECT_LE(normInf, lInf) << comment;
 }
 
-std::string findModel(const std::string &model_name)
-{
-    return findDataFile("vision/classification/squeezenet/model/" + model_name + ".onnx", false);
+std::string findModel(const std::string &model_name) {
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:57]: I don't know what findDataFile does. M.b. it is large...

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 7]
```diff
@@ -12,172 +12,309 @@
...
-                       const cv::Scalar& std) {
-    int new_h = 224;
-    int new_w = 224;
-    cv::Mat tmp, nmat, cvt;
-    cv::resize(src, dst, cv::Size(new_w, new_h));
-    dst.convertTo(cvt, CV_32F, 1.f / 255);
-    nmat = cvt - mean;
-    tmp = nmat / std;
-    dst.create(cv::Size(new_w, new_h * src.channels()), CV_32F);
+inline void toCHW(const cv::Mat& src, cv::Mat& dst) {
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:61]: ?

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 8]
```diff
@@ -12,172 +12,309 @@
...
+}
 
-TEST(ONNX, Infer)
-{
-    cv::Mat in_mat1, out_gapi, out_onnx;
-    std::string model_path = findModel("squeezenet1.0-9");
-    // NOTE: All tests chek "random" image
-    // Ideally it should be a real image
-    in_mat1 = initMatrixRandU(CV_8UC3, cv::Size{640, 480});
+inline std::vector<const char*> getCharNames(const std::vector<std::string>& names) {
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:83]: ?

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 9]
```diff
@@ -12,172 +12,309 @@
...
+    std::vector<const char*> out_vec;
+    for (const auto& el : names) {
+        out_vec.push_back(el.data());
+    }
+    return out_vec;
+}
 
-    cv::Scalar mean = { 0.485, 0.456, 0.406 };
-    cv::Scalar std  = { 0.229, 0.224, 0.225 };
+inline void match_out(const cv::Mat& in, cv::Mat& out) {
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: ?

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 10]
```diff
@@ -12,172 +12,309 @@
...
+    GAPI_Assert(in.depth() == CV_32F);
+    GAPI_Assert(in.size == out.size);
+    const float* inptr = in.ptr<float>();
+    float* optr = out.ptr<float>();
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:95]: No. optr can't.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 11]
```diff
@@ -12,172 +12,309 @@
...
+    const int size = in.total();
+    for (int i = 0; i < size; ++i) {
+        optr[i] = inptr[i];
+    }
+}
 
-    // ONNX_API code
-    InferONNX(model_path, in_mat1, out_onnx, mean, std);
+void remap_yolo(const std::unordered_map<std::string, cv::Mat> &onnx,
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 12]
```diff
@@ -12,172 +12,309 @@
...
+    GAPI_Assert(gapi.size() == 1u);
+    // Result from Run method
+    const cv::Mat& in = onnx.begin()->second;
+    // Configured output
+    cv::Mat& out = gapi.begin()->second;
+    // Simple copy
+    match_out(in, out);
+}
+
+void remap_ssd_ports(const std::unordered_map<std::string, cv::Mat> &onnx,
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 13]
```diff
@@ -12,172 +12,309 @@
...
+private:
+    Ort::Env env{nullptr};
+    Ort::MemoryInfo memory_info{nullptr};
+    Ort::AllocatorWithDefaultOptions allocator;
+    Ort::SessionOptions session_options;
+    Ort::Session session{nullptr};
+
+    std::vector<int64_t> input_node_dims;
+    std::vector<std::string> in_node_names;
+    std::vector<std::string> out_node_names;
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:231]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 14]
```diff
@@ -12,172 +12,309 @@
...
+        num_out = session.GetOutputCount();
+        in_node_names.clear();
+        out_node_names.clear();
+        // Inputs Run params
+        std::vector<Ort::Value> in_tensors;
+        for(size_t i = 0; i < num_in; ++i) {
+            char* in_node_name_p = session.GetInputName(i, allocator);
+            in_node_names.push_back(std::string(in_node_name_p));
+            allocator.Free(in_node_name_p);
+            input_node_dims = toORT(ins[i].size);
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: It isn't the same. toORT is more flexible.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 15]
```diff
@@ -12,172 +12,309 @@
...
+            cv::Mat mt(std::vector<int>(shape.begin(), shape.end()), toCV(type),
+                       reinterpret_cast<void*>(result[i].GetTensorMutableData<uint8_t*>()));
+            mt.copyTo(outs[i]);
+        }
+    }
+    // One input/output overload
+    template<typename T>
+    void infer(const cv::Mat& in, cv::Mat& out) {
+        std::vector<cv::Mat> result;
+        infer<T>({in}, result);
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:205]: No

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 16]
```diff
@@ -12,172 +12,309 @@
...
+        GAPI_Assert(!out_gapi.empty() && !out_onnx.empty());
+        ASSERT_EQ(out_gapi.size(), out_onnx.size());
+        const auto size = out_gapi.size();
+        for (size_t i = 0; i < size; ++i) {
+            normAssert(out_onnx[i], out_gapi[i], "Test outputs");
+        }
+    }
+
+    void useModel(const std::string& model_name) {
+        model_path = findModel(model_name);
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:220]: infer() uses model_path. All tests use model_path. We can put `model_path = findModel(model_name);` in every test instead this function.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 17]
```diff
@@ -12,172 +12,309 @@
...
+    std::vector<std::string> in_node_names;
+    std::vector<std::string> out_node_names;
+};
+
+class ONNXClassificationTest : public ONNXtest {
+public:
+    cv::Scalar mean, std;
+    virtual void SetUp() {
+        mean = { 0.485, 0.456, 0.406 };
+        std  = { 0.229, 0.224, 0.225 };
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Ok. We will return this if classifications require different values.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 18]
```diff
@@ -12,172 +12,309 @@
...
+    }
+
+    void preprocess(const cv::Mat& src, cv::Mat& dst) {
+        const int new_h = 224;
+        const int new_w = 224;
+        cv::Mat tmp, nmat, cvt;
+        cv::resize(src, dst, cv::Size(new_w, new_h));
+        dst.convertTo(cvt, CV_32F, 1.f / 255);
+        nmat = cvt - mean;
+        tmp = nmat / std;
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Yes.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 19]
```diff
@@ -12,172 +12,309 @@
...
+    virtual void SetUp() {
+        mean = { 0.485, 0.456, 0.406 };
+        std  = { 0.229, 0.224, 0.225 };
+    }
+
+    void preprocess(const cv::Mat& src, cv::Mat& dst) {
+        const int new_h = 224;
+        const int new_w = 224;
+        cv::Mat tmp, nmat, cvt;
+        cv::resize(src, dst, cv::Size(new_w, new_h));
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Ok. So clearer

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 20]
```diff
@@ -12,172 +12,309 @@
...
+        dst = dst.reshape(1, {1, 3, new_h, new_w});
+    }
+};
+
+class ONNXGRayScaleTest : public ONNXtest {
+public:
+    void preprocess(const cv::Mat& src, cv::Mat& dst) {
+        const int new_h = 64;
+        const int new_w = 64;
+        cv::Mat csc, rsc, cvt;
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: cvc and rsz

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 21]
```diff
@@ -12,172 +12,309 @@
...
+inline std::vector<int64_t> toORT(const cv::MatSize &sz) {
+    return cv::to_own<int64_t>(sz);
+}
 
-TEST(ONNX, Infer)
-{
-    cv::Mat in_mat1, out_gapi, out_onnx;
-    std::string model_path = findModel("squeezenet1.0-9");
-    // NOTE: All tests chek "random" image
-    // Ideally it should be a real image
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:141]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 22]
```diff
@@ -264,15 +381,82 @@ TEST(ONNX, Infer2ROIList)
     comp.apply(cv::gin(in_mat1, rois),
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:390]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 23]
```diff
@@ -225,35 +353,24 @@ TEST(ONNX, InferROIList)
...
     };
-    std::vector<cv::Mat> out_gapi;
-    std::vector<cv::Mat> out_onnx(rois.size());
     // ONNX_API code
+    out_onnx.resize(rois.size());
     for (size_t i = 0; i < rois.size(); ++i) {
-        InferONNX(model_path, in_mat1(rois[i]), out_onnx[i], mean, std);
+        cv::Mat roi_mat;
+        preprocess(in_mat1(rois[i]), roi_mat);
+        infer<float>(roi_mat, out_onnx[i]);
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:379]: Yes. It is how it was intended. But we must to create fixture for ROI tests. But i don't want to touch classification fixture now. And i don't know what tests will be added with other features (multiple inputs exmpl.). I want to keep it simple.

[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:379]: I'll come back to it with next PR with multiple input check.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 1 above)
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:156]: Fixed. But ...

Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:156]: Nothing

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 2 above)
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 3 above)
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 4 above)
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Fixed.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 24]
```diff
@@ -12,172 +12,309 @@
...
+        // Copy outputs
+        GAPI_Assert(result.size() == num_out);
+        outs.resize(num_out);
+        for (size_t i = 0; i < num_out; ++i) {
+            const auto info = result[i].GetTensorTypeAndShapeInfo();
+            const auto shape = info.GetShape();
+            const auto type = info.GetElementType();
+            cv::Mat mt(std::vector<int>(shape.begin(), shape.end()), toCV(type),
+                       reinterpret_cast<void*>(result[i].GetTensorMutableData<uint8_t*>()));
+            mt.copyTo(outs[i]);
```
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:198]: No. We need data copy here.

**Comment by mpashchenkov** [comment]
@alalek Please add [tinyyolov2-8.onnx](https://github.com/onnx/models/tree/master/vision/object_detection_segmentation/tiny-yolov2) for bb custom ONNX.

I added it to PR description.

One test throws: Can't find required data file: vision/object_detection_segmentation/tiny-yolov2/model/tinyyolov2-8.onnx

File: modules/gapi/src/backends/onnx/gonnxbackend.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/src/backends/onnx/gonnxbackend.cpp:177]: OK, as you assume

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:61]: let it be

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:83]: let it be

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (mpashchenkov) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:57]: Ok. It is recommendation.

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:205]: Yes, I see now, too, sorry

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 24 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:198]: OK, I see

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 23 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:379]: Agree. Don't forget :)

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp:220]: Yes, we should preserve the function outside the class as there can be tests which are not created via fixture. I agree

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: I mixed up const pointers, sorry. Everything is OK here. You can add `* const` if you wish

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: let it be

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (OrestChura) [reply modules/gapi/test/infer/gapi_infer_onnx_test.cpp]: Good

**Review by OrestChura — APPROVED** [review]
Thanks!

**Comment by mpashchenkov** [comment]
@dmatveev Last words about it (one model is missing yet).

File: modules/gapi/test/infer/gapi_infer_onnx_test.cpp [hunk 25]
```diff
@@ -12,267 +12,462 @@
...
-    cv::Scalar std  = { 0.229, 0.224, 0.225 };
+inline void copyToOut(const cv::Mat& in, cv::Mat& out) {
+    GAPI_Assert(in.depth() == CV_32F);
+    GAPI_Assert(in.size == out.size);
+    const float* const inptr = in.ptr<float>();
+    float* const optr = out.ptr<float>();
+    const int size = in.total();
+    for (int i = 0; i < size; ++i) {
+        optr[i] = inptr[i];
+    }
```
Comment (dmatveev) [inline modules/gapi/test/infer/gapi_infer_onnx_test.cpp:99]: `Mat::copyTo()` is enough, assuming sizes are equal.

**Review by dmatveev — APPROVED** [review]
"Approved with comments" 👍 

Please fix the stuff in the follow-up MR.

**Pull request merged**